allennlp/data/fields/text_field.py

```diff
@@ -37,7 +37,7 @@ class TextField(SequenceField[Dict[str, torch.Tensor]]):
     def __init__(self, tokens: List[Token], token_indexers: Dict[str, TokenIndexer]) -> None:
         self.tokens = tokens
         self._token_indexers = token_indexers
-        self._indexed_tokens: Optional[Dict[str, TokenList]] = None
+        self._indexed_tokens: Optional[Dict[str, Any]] = None
```

Comment:
so I was actually imagining something like where a BPE indexer would generate two top-level arrays

`f"{indexer_name}-byte-pairs"` and `f"{indexer_name}-offsets"`

rather than a dict of them. that's clunkier in some ways but again it avoids the `isinstance` checks and having to have two sorts of logic.

thoughts?

Reply:
I'm not sure I understand?  The token indexer is returning a dict of lists following the type hint, see the test stub `DictReturningTokenIndexer` in `tests/data/fields/text_field_test.py`.

Reply:
right, I was thinking that `TextField.index` would *merge* the returned dicts, producing a one-level dict of the indices.

that is, if one indexer returned `{'byte-pairs': [..], 'offsets': [...]}` and another `{'tokens': [...]}`, it would produce a dict with 3 keys: byte-pairs, offsets, and tokens

whereas this change would make it return a dict with two keys, one of which held some indices and one of which held another dict?

Reply:
I agree with Joel here - this will work seamlessly with the changes you've already made to the `TextFieldEmbedder` which allow mis-matching keys, whilst reducing the complexity in the indexing code. The case that a `TokenIndexer` generates multiple arrays is rare enough that if you want to do this, you'll just have to use the `indexer_to_embedder_map` argument in the `TextFieldEmbedder` to map the arrays to the embedder. 

The only downside with creating a flat map is that the 1-1 map between embedders and indexer names is now not enforceable - we should add some documentation that the key passed to `tokens_to_indices` really should be used as a key in isolation, even if it is then used to create auxilary keys for other return arrays.

e.g if you have a byte_pair_embedder called `bytes`, we really need the returned array to have exactly `bytes` as a key: `{bytes: [....],  bytes-offsets: [....]}`, but we can't actually enforce this if users were to implement their own indexers which return multiple arrays.

I think this is ok though, because in reality I don't imagine that there are many people implementing new indexers and if they do, they know what they are doing.

Reply:
We need to maintain the mapping between the token indexers in `self._token_indexers` and the indexed tokens in `self._indexed_tokens` to call `indexer.get_padding_lengths` (in `get_padding_lengths`) and `indexer.pad_token_sequence` in `as_tensor`).

If we flatten the `self._indexed_tokens` dict as suggested then we need to introduce another dict to maintain this mapping.  An alternate implementation I've chosen encodes this as the type of `self._indexed_tokens[indexer_name]`.

Reply:
Also -- these are just the details of an internal private attribute `_indexed_tokens` -- @markn we still need to provide a `indexer_to_embedder_map` in `TextFieldEmbedder` independent of how this is implemented.

Reply:
Right - I was considering the case where we also return tensors in the nested format. Maintaining another mapping seems cleaner to me if it avoids the type checks.